### PR TITLE
use safe toString for array assetion

### LINKED ISSUE #15421: addObserverForContentKey performance in 2.13.3
Hi,

I ran into an issue with debug builds taking ages to run when using test data with somewhat large arrays. I tracked it down to `addObserverForContentKey`, which I believe was already an issue earlier #3156

https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/mixins/array.js#L735
```
 assert(`When using @each to observe the array ${content}, the array must return an object`, typeof item === 'object');
```
This is not a problem if content is an ArrayProxy, but a native array will do Array.toString() on this and will blow things up.

I know this is fixed in release builds, but not being able to test my app with large datasets is not good at all.

### PR DESCRIPTION
**Pull request by bekzod**
fixes #15421

### CODE REVIEW

**Comment by rwjblue** [comment]
As I mentioned in #15421 the changes to the assertion in `addObserverForContentKey` are not required (as of 2.14 and higher).

The other changes (adding `meta` to `propertyDidChange`) seem quite unrelated (though perhaps good on their own?).  Mind splitting that out and submitting that separately?

**Comment by bekzod** [comment]
@rwjblue won't `content` blow up if it is `[Object.create(null), {}]` because that is reason I used `ember-utils`'s `toString`

**Comment by rwjblue** [comment]
@bekzod - Yes, but only if the assertion would already fail (e.g. not the original scenario from #15421).  I think moving to using the custom `toString` from `ember-utils` is totally fine for that part, but we definitely do not want to eagerly calculate it (which the PR does at the moment).

**Comment by rwjblue** [comment]
I'm going to reopen so we can update that.

File: packages/ember-runtime/lib/mixins/array.js [hunk 1]
```diff
@@ -150,16 +151,16 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
 
     let normalStartIdx = startIdx < 0 ? previousLength + startIdx : startIdx;
     if (cache.firstObject !== undefined && normalStartIdx === 0) {
-      propertyWillChange(array, 'firstObject');
-      propertyDidChange(array, 'firstObject');
+      propertyWillChange(array, 'firstObject', meta);
```
Comment (rwjblue) [inline packages/ember-runtime/lib/mixins/array.js]: Please migrate this change to a separate / targeted PR.

File: packages/ember-runtime/lib/mixins/array.js [hunk 2]
```diff
@@ -150,16 +151,16 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
...
+      propertyDidChange(array, 'firstObject', meta);
     }
 
     if (cache.lastObject !== undefined) {
       let previousLastIndex = previousLength - 1;
       let lastAffectedIndex = normalStartIdx + removedAmount;
       if (previousLastIndex < lastAffectedIndex) {
-        propertyWillChange(array, 'lastObject');
-        propertyDidChange(array, 'lastObject');
+        propertyWillChange(array, 'lastObject', meta);
```
Comment (rwjblue) [inline packages/ember-runtime/lib/mixins/array.js]: Please migrate this change to a separate / targeted PR.

File: packages/ember-runtime/lib/mixins/array.js [hunk 3]
```diff
@@ -740,10 +741,13 @@ EachProxy.prototype = {
 };
 
 function addObserverForContentKey(content, keyName, proxy, idx, loc) {
+  let arrayString;
+  if (DEBUG) { arrayString = toString(content); }
```
Comment (rwjblue) [inline packages/ember-runtime/lib/mixins/array.js]: Please remove the `arrayString` variable, and update the assertion message to call `${toString(content)}` instead.

**Review by rwjblue — CHANGES_REQUESTED** [review]

**Comment by bekzod** [comment]
@rwjblue done

**Review by rwjblue — APPROVED** [review]

**Comment by rwjblue** [comment]
Awesome, thank you!!

**Pull request merged**